allennlp/modules/conditional_random_field.py

```diff
@@ -19,15 +76,29 @@ class ConditionalRandomField(torch.nn.Module):
     ----------
     num_tags : int, required
         The number of tags.
+    constraints : List[Tuple[int, int]], optional (default: None)
+        An optional list of allowed transitions (from_tag_id, to_tag_id).
+        These are applied to ``viterbi_tags()`` but do not affect ``forward()``.
     """
     def __init__(self,
-                 num_tags: int) -> None:
+                 num_tags: int,
+                 constraints: List[Tuple[int, int]] = None) -> None:
         super().__init__()
         self.num_tags = num_tags
 
         # transitions[i, j] is the logit for transitioning from state i to state j.
         self.transitions = torch.nn.Parameter(torch.Tensor(num_tags, num_tags))
 
+        # _constraint_mask indicates valid transitions (based on supplied constraints).
+        if constraints is None:
+            self._constraint_mask = None
+        else:
+            constraint_mask = torch.Tensor(num_tags, num_tags).fill_(0.)
+            for i, j in constraints:
+                constraint_mask[i, j] = 1.
+
+            self._constraint_mask = torch.nn.Parameter(constraint_mask, requires_grad=False)
```

Comment:
This seems like it should be fine - the only other way I could think of would be to allow `viterbi_tags` to take the transition matrix as a parameter, so you could just pass it in at runtime having created it in the CrfTagger. But this looks fine.